File: apps/ios/Data/Sources/Data/ReceiptsRepository.swift
import Foundation
import PowerSync
import Supabase
import Domain

// Read/write facade for receipt scans (P2.5): receipt_scans table.
// Mirrors apps/web/src/receipts/scan.ts (saveScan, updateScanDraft, linkScan).
// Mirrors apps/android/data/.../repository/ReceiptsRepository.kt.

public struct ReceiptScanRow: Sendable {
    public let id: String
    public let userId: String
    public let source: String
    public let engine: String
    public let merchant: String?
    public let occurredAt: String?
    public let currency: String?
    public let subtotal: Int64?
    public let tax: Int64?
    public let serviceCharge: Int64?
    public let tip: Int64?
    public let discount: Int64?
    public let total: Int64?
    public let confidence: Int64?
    public let rawText: String?
    public let parsedJson: String?
    public let transactionId: String?
    public let expenseId: String?
    public let imagePath: String?
    public let createdAt: String
    public let updatedAt: String
}

public struct SaveScanInput: Sendable {
    public let source: String
    public let engine: String
    public let merchant: String?
    public let occurredAt: String?
    public let currency: String?
    public let subtotal: Int64?
    public let tax: Int64?
    public let serviceCharge: Int64?
    public let tip: Int64?
    public let discount: Int64?
    public let total: Int64?
    public let confidence: Int64?
    public let rawText: String?
    public let parsedJson: String?

    public init(
        source: String,
        engine: String,
        merchant: String? = nil,
        occurredAt: String? = nil,
        currency: String? = nil,
        subtotal: Int64? = nil,
        tax: Int64? = nil,
        serviceCharge: Int64? = nil,
        tip: Int64? = nil,
        discount: Int64? = nil,
        total: Int64? = nil,
        confidence: Int64? = nil,
        rawText: String? = nil,
        parsedJson: String? = nil
    ) {
        self.source = source
        self.engine = engine
        self.merchant = merchant
        self.occurredAt = occurredAt
        self.currency = currency
        self.subtotal = subtotal
        self.tax = tax
        self.serviceCharge = serviceCharge
        self.tip = tip
        self.discount = discount
        self.total = total
        self.confidence = confidence
        self.rawText = rawText
        self.parsedJson = parsedJson
    }
}

public struct UpdateScanDraftInput: Sendable {
    public let engine: String
    public let merchant: String?
    public let occurredAt: String?
    public let currency: String?
    public let subtotal: Int64?
    public let tax: Int64?
    public let serviceCharge: Int64?
    public let tip: Int64?
    public let discount: Int64?
    public let total: Int64?
    public let confidence: Int64?
    public let parsedJson: String?

    public init(
        engine: String,
        merchant: String? = nil,
        occurredAt: String? = nil,
        currency: String? = nil,
        subtotal: Int64? = nil,
        tax: Int64? = nil,
        serviceCharge: Int64? = nil,
        tip: Int64? = nil,
        discount: Int64? = nil,
        total: Int64? = nil,
        confidence: Int64? = nil,
        parsedJson: String? = nil
    ) {
        self.engine = engine
        self.merchant = merchant
        self.occurredAt = occurredAt
        self.currency = currency
        self.subtotal = subtotal
        self.tax = tax
        self.serviceCharge = serviceCharge
        self.tip = tip
        self.discount = discount
        self.total = total
        self.confidence = confidence
        self.parsedJson = parsedJson
    }
}

private func currentIsoString() -> String {
    ISO8601DateFormatter().string(from: Date())
}

public final class ReceiptsRepository: @unchecked Sendable {
    private let db: PowerSyncDatabaseProtocol
    private let getUserId: @Sendable () -> String
    /// Only the AI fallback needs the network. Everything else on this
    /// repository is local — which is the point of the feature, and is why the
    /// client arrives as a dependency rather than being reached for globally.
    private let client: SupabaseClient

    public init(
        db: PowerSyncDatabaseProtocol,
        getUserId: @escaping @Sendable () -> String,
        client: SupabaseClient
    ) {
        self.db = db
        self.getUserId = getUserId
        self.client = client
    }

    /**
     Send the photo to the `receipt-scan` edge function and map its reply.

     **The only code path in this feature where the image leaves the device**,
     and it is reached only from an explicit "Improve with AI" tap. The scan
     pipeline never calls it.

     The MAPPING is not here — `aiReceiptDraft` in Domain does that under 18
     vectors, because it decides money from untrusted input. This does the two
     things a repository should: the call, and turning a failure into something
     the UI can act on. `quotaExceeded` is separated from every other error
     because it is the one failure with a next step: web shows the upgrade path
     for it and only for it.
     */
    public func aiParseReceipt(
        base64: String,
        mediaType: String,
        currencyHint: String,
        today: String,
        rawText: String? = nil
    ) async throws -> ReceiptDraft {
        let body: [String: String] = [
            "image": base64,
            "mediaType": mediaType,
            "currencyHint": currencyHint,
            "today": today,
        ]
        let data: Foundation.Data
        do {
            // The raw-bytes overload, matching AssistantRepository: the reply is
            // a dynamic receipt shape, and a Decodable struct here would turn
            // one unexpected field into a total failure.
            data = try await client.functions.invoke(
                "receipt-scan",
                options: FunctionInvokeOptions(body: body)
            ) { raw, _ in raw }
        } catch let error as FunctionsError {
            // A non-2xx carries the real reason in its body, and the function
            // always answers with `{ error, code }` — the same unwrapping web's
            // `edgeFnMessage()` does, for the same reason. The code matters as
            // much as the message here.
            if case let .httpError(code, body) = error {
                let parsed = (try? JSONSerialization.jsonObject(with: body)) as? [String: Any]
                throw AiScanError(
                    message: (parsed?["error"] as? String) ?? aiScanStatusMessage(code),
                    quotaExceeded: (parsed?["code"] as? String) == "quota_exceeded"
                )
            }
            throw AiScanError(message: error.localizedDescription)
        } catch {
            throw AiScanError(message: error.localizedDescription)
        }

        guard let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            throw AiScanError(message: aiScanEmpty)
        }
        if let message = json["error"] as? String {
            throw AiScanError(
                message: message,
                quotaExceeded: (json["code"] as? String) == "quota_exceeded"
            )
        }
        guard let receipt = json["receipt"] as? [String: Any] else {
            throw AiScanError(message: aiScanEmpty)
        }
        return aiReceiptDraft(
            receipt: aiReceipt(from: receipt),
            currencyHint: currencyHint,
            rawText: rawText
        )
    }

    public func saveScan(_ input: SaveScanInput) async throws -> String {
        let id = UUID().canonicalString
        let ts = currentIsoString()
        let userId = getUserId()
        let rawTextCapped = input.rawText.map { String($0.prefix(8000)) }

        try await db.execute(
            sql: """
            INSERT INTO receipt_scans (
                id, user_id, source, engine, merchant, occurred_at, currency,
                subtotal, tax, service_charge, tip, discount, total, confidence,
                raw_text, parsed_json, transaction_id, expense_id, image_path,
                created_at, updated_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            parameters: [
                id, userId, input.source, input.engine, input.merchant, input.occurredAt, input.currency,
                input.subtotal, input.tax, input.serviceCharge, input.tip, input.discount, input.total, input.confidence,
                rawTextCapped, input.parsedJson, nil, nil, nil,
                ts, ts
            ]
        )
        return id
    }

    public func updateScanDraft(scanId: String, input: UpdateScanDraftInput) async throws {
        let ts = currentIsoString()
        try await db.execute(
            sql: """
            UPDATE receipt_scans SET
                engine = ?, merchant = ?, occurred_at = ?, currency = ?,
                subtotal = ?, tax = ?, service_charge = ?, tip = ?, discount = ?,
                total = ?, confidence = ?, parsed_json = ?, updated_at = ?
            WHERE id = ? AND deleted_at IS NULL
            """,
            parameters: [
                input.engine, input.merchant, input.occurredAt, input.currency,
                input.subtotal, input.tax, input.serviceCharge, input.tip, input.discount,
                input.total, input.confidence, input.parsedJson, ts,
                scanId
            ]
        )
    }

    public func linkScan(scanId: String, transactionId: String? = nil, expenseId: String? = nil) async throws {
        let ts = currentIsoString()
        var sets: [String] = []
        var params: [Sendable?] = []

        if let tId = transactionId {
            sets.append("transaction_id = ?")
            params.append(tId)
        }
        if let eId = expenseId {
            sets.append("expense_id = ?")
            params.append(eId)
        }
        if sets.isEmpty { return }

        sets.append("updated_at = ?")
        params.append(ts)
        params.append(scanId)

        try await db.execute(
            sql: "UPDATE receipt_scans SET \(sets.joined(separator: ", ")) WHERE id = ? AND deleted_at IS NULL",
            parameters: params
        )
    }

    public func get(scanId: String) async throws -> ReceiptScanRow? {
        try await db.getOptional(
            sql: "SELECT * FROM receipt_scans WHERE id = ? AND deleted_at IS NULL",
            parameters: [scanId]
        ) { cursor in
            ReceiptScanRow(
                id: try cursor.getString(name: "id"),
                userId: try cursor.getString(name: "user_id"),
                source: try cursor.getString(name: "source"),
                engine: try cursor.getString(name: "engine"),
                merchant: try cursor.getStringOptional(name: "merchant"),
                occurredAt: try cursor.getStringOptional(name: "occurred_at"),
                currency: try cursor.getStringOptional(name: "currency"),
                subtotal: try cursor.getInt64Optional(name: "subtotal"),
                tax: try cursor.getInt64Optional(name: "tax"),
                serviceCharge: try cursor.getInt64Optional(name: "service_charge"),
                tip: try cursor.getInt64Optional(name: "tip"),
                discount: try cursor.getInt64Optional(name: "discount"),
                total: try cursor.getInt64Optional(name: "total"),
                confidence: try cursor.getInt64Optional(name: "confidence"),
                rawText: try cursor.getStringOptional(name: "raw_text"),
                parsedJson: try cursor.getStringOptional(name: "parsed_json"),
                transactionId: try cursor.getStringOptional(name: "transaction_id"),
                expenseId: try cursor.getStringOptional(name: "expense_id"),
                imagePath: try cursor.getStringOptional(name: "image_path"),
                createdAt: try cursor.getString(name: "created_at"),
                updatedAt: try cursor.getString(name: "updated_at")
            )
        }
    }

    public func list(limit: Int = 50) async throws -> [ReceiptScanRow] {
        try await db.getAll(
            sql: "SELECT * FROM receipt_scans WHERE deleted_at IS NULL ORDER BY created_at DESC LIMIT ?",
            parameters: [limit]
        ) { cursor in
            ReceiptScanRow(
                id: try cursor.getString(name: "id"),
                userId: try cursor.getString(name: "user_id"),
                source: try cursor.getString(name: "source"),
                engine: try cursor.getString(name: "engine"),
                merchant: try cursor.getStringOptional(name: "merchant"),
                occurredAt: try cursor.getStringOptional(name: "occurred_at"),
                currency: try cursor.getStringOptional(name: "currency"),
                subtotal: try cursor.getInt64Optional(name: "subtotal"),
                tax: try cursor.getInt64Optional(name: "tax"),
                serviceCharge: try cursor.getInt64Optional(name: "service_charge"),
                tip: try cursor.getInt64Optional(name: "tip"),
                discount: try cursor.getInt64Optional(name: "discount"),
                total: try cursor.getInt64Optional(name: "total"),
                confidence: try cursor.getInt64Optional(name: "confidence"),
                rawText: try cursor.getStringOptional(name: "raw_text"),
                parsedJson: try cursor.getStringOptional(name: "parsed_json"),
                transactionId: try cursor.getStringOptional(name: "transaction_id"),
                expenseId: try cursor.getStringOptional(name: "expense_id"),
                imagePath: try cursor.getStringOptional(name: "image_path"),
                createdAt: try cursor.getString(name: "created_at"),
                updatedAt: try cursor.getString(name: "updated_at")
            )
        }
    }

    /**
     Transaction ids that came from a receipt photo, live.

     Web's `useScannedTransactionIds()` (apps/web/src/splits/hooks.ts) reads
     exactly this and the Transactions list uses it to draw the "Scanned" pill.
     Neither phone ever queried `receipt_scans.transaction_id`, so a scanned
     transaction looked like any other hand-typed one.

     A whole-column watch rather than a per-row lookup: the list needs the
     answer for up to 200 rows at once, and this table has one row per scan —
     two orders of magnitude smaller than the ledger it annotates.
     */
    public func watchScannedTransactionIds() throws -> AsyncThrowingStream<Set<String>, Error> {
        let rows: AsyncThrowingStream<[String], Error> = try db.watch(
            sql: "SELECT transaction_id FROM receipt_scans WHERE transaction_id IS NOT NULL AND deleted_at IS NULL",
            parameters: [],
            mapper: { cursor in try cursor.getString(name: "transaction_id") }
        )
        return AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    for try await batch in rows { continuation.yield(Set(batch)) }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { @Sendable _ in task.cancel() }
        }
    }

    public func delete(scanId: String) async throws {
        let ts = currentIsoString()
        try await db.execute(
            sql: "UPDATE receipt_scans SET deleted_at = ?, updated_at = ? WHERE id = ?",
            parameters: [ts, ts, scanId]
        )
    }
}

/**
 A failed AI read.

 `quotaExceeded` is separate because it is the only failure the user can do
 something about: web shows the upgrade path for it and a plain message for
 everything else.
 */
public struct AiScanError: LocalizedError {
    public let message: String
    public let quotaExceeded: Bool
    public init(message: String, quotaExceeded: Bool = false) {
        self.message = message
        self.quotaExceeded = quotaExceeded
    }
    public var errorDescription: String? { message }
}

private let aiScanFailure = "Couldn't reach the scanner. Check your connection."
private let aiScanEmpty = "The scan came back empty. Try a clearer photo."

private func aiScanStatusMessage(_ code: Int) -> String {
    code == 401 ? "Please sign in to use AI receipt reading." : aiScanFailure
}

/**
 The edge function's `receipt` object, as Domain's input type.

 Deliberately tolerant: every field is optional and a wrong TYPE reads as absent
 rather than throwing. A model reply is untrusted input, and a strict decoder
 here would turn one odd field into a total failure where web would have shown
 the user a partial draft they could fix.

 `as? NSNumber` is the `typeof x === "number"` guard: a model that returns
 `"12.50"` as a STRING is rejected, and the line dropped, exactly as on web.
 */
private func aiText(_ d: [String: Any], _ key: String) -> String? {
    if let s = d[key] as? String { return s }
    if let n = d[key] as? NSNumber { return n.stringValue }
    return nil
}

private func aiNumber(_ d: [String: Any], _ key: String) -> Double? {
    (d[key] as? NSNumber)?.doubleValue
}

private func aiReceipt(from d: [String: Any]) -> AiReceipt {
    AiReceipt(
        merchant: aiText(d, "merchant"),
        date: aiText(d, "date"),
        currency: aiText(d, "currency"),
        total: aiNumber(d, "total"),
        confidence: aiNumber(d, "confidence"),
        lines: ((d["lines"] as? [Any]) ?? []).compactMap { raw in
            guard let o = raw as? [String: Any] else { return nil }
            return AiLine(
                kind: aiText(o, "kind"),
                description: aiText(o, "description"),
                quantity: aiNumber(o, "quantity"),
                unit: aiText(o, "unit"),
                // The wire name is snake_case; Domain's field is not.
                // `receipts-ai.json` carries the WIRE name so the corpus pins
                // it and both platforms' registrations must agree.
                unitPrice: aiNumber(o, "unit_price"),
                amount: aiNumber(o, "amount")
            )
        }
    )
}

File: apps/android/data/src/main/kotlin/com/sanvya/app/data/repository/ReceiptsRepository.kt
package com.sanvya.app.data.repository

/**
 * Read/write facade for receipt scans (P2.5): receipt_scans table.
 * Mirrors apps/web/src/receipts/scan.ts (saveScan, updateScanDraft, linkScan).
 *
 * Table columns confirmed against PocketCareSchema.kt (P2.1) and
 * supabase/migrations/0001_init.sql.
 */

import com.powersync.PowerSyncDatabase
import com.powersync.db.getLongOptional
import com.powersync.db.getString
import com.powersync.db.getStringOptional
import com.sanvya.app.domain.receipts.AiLine
import com.sanvya.app.domain.receipts.AiReceipt
import com.sanvya.app.domain.receipts.ReceiptDraft
import com.sanvya.app.domain.receipts.aiReceiptDraft
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.exceptions.RestException
import io.github.jan.supabase.functions.functions
import io.ktor.client.statement.bodyAsText
import io.ktor.http.Headers
import io.ktor.http.HttpHeaders
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import java.time.Instant
import java.util.UUID

data class ReceiptScanRow(
    val id: String,
    val userId: String,
    val source: String,
    val engine: String,
    val merchant: String?,
    val occurredAt: String?,
    val currency: String?,
    val subtotal: Long?,
    val tax: Long?,
    val serviceCharge: Long?,
    val tip: Long?,
    val discount: Long?,
    val total: Long?,
    val confidence: Long?,
    val rawText: String?,
    val parsedJson: String?,
    val transactionId: String?,
    val expenseId: String?,
    val imagePath: String?,
    val createdAt: String,
    val updatedAt: String,
)

data class SaveScanInput(
    val source: String,
    val engine: String,
    val merchant: String? = null,
    val occurredAt: String? = null,
    val currency: String? = null,
    val subtotal: Long? = null,
    val tax: Long? = null,
    val serviceCharge: Long? = null,
    val tip: Long? = null,
    val discount: Long? = null,
    val total: Long? = null,
    val confidence: Long? = null,
    val rawText: String? = null,
    val parsedJson: String? = null,
)

data class UpdateScanDraftInput(
    val engine: String,
    val merchant: String? = null,
    val occurredAt: String? = null,
    val currency: String? = null,
    val subtotal: Long? = null,
    val tax: Long? = null,
    val serviceCharge: Long? = null,
    val tip: Long? = null,
    val discount: Long? = null,
    val total: Long? = null,
    val confidence: Long? = null,
    val parsedJson: String? = null,
)

class ReceiptsRepository(
    private val db: PowerSyncDatabase,
    private val getUserId: () -> String,
    /**
     * Only the AI fallback needs the network. Everything else on this
     * repository is local -- which is the point of the feature, and is why the
     * client arrives as a dependency rather than being reached for globally.
     */
    private val client: SupabaseClient,
) {
    /**
     * Send the photo to the `receipt-scan` edge function and map its reply.
     *
     * **The only code path in this feature where the image leaves the device**,
     * and it is reached only from an explicit "Improve with AI" tap. The scan
     * pipeline never calls it.
     *
     * The MAPPING is not here -- `aiReceiptDraft` in Domain does that under 18
     * vectors, because it decides money from untrusted input. This does the two
     * things a repository should: the call, and turning a failure into
     * something the UI can act on. `quotaExceeded` is separated from every
     * other error because it is the one failure with a next step: web shows the
     * upgrade path for it and only for it.
     */
    suspend fun aiParseReceipt(
        base64: String,
        mediaType: String,
        currencyHint: String,
        today: String,
        rawText: String? = null,
    ): ReceiptDraft {
        val body = buildJsonObject {
            put("image", base64)
            put("mediaType", mediaType)
            put("currencyHint", currencyHint)
            put("today", today)
        }
        val json = try {
            val response = client.functions.invoke(
                function = "receipt-scan",
                body = body,
                headers = Headers.build { append(HttpHeaders.ContentType, "application/json") },
            )
            Json.parseToJsonElement(response.bodyAsText()).jsonObject
        } catch (e: RestException) {
            // supabase-kt collapses a non-2xx into an exception whose message is
            // the raw body, and the function always answers with `{ error }` --
            // the same unwrapping web's `edgeFnMessage()` does, for the same
            // reason. The code matters as much as the message here.
            val parsed = runCatching { Json.parseToJsonElement(e.message.orEmpty()).jsonObject }.getOrNull()
            throw AiScanError(
                message = parsed?.get("error")?.jsonPrimitive?.contentOrNull ?: AI_SCAN_FAILURE,
                quotaExceeded = parsed?.get("code")?.jsonPrimitive?.contentOrNull == "quota_exceeded",
            )
        } catch (e: Exception) {
            throw AiScanError(e.message ?: AI_SCAN_FAILURE)
        }

        json["error"]?.jsonPrimitive?.contentOrNull?.let { message ->
            throw AiScanError(
                message = message,
                quotaExceeded = json["code"]?.jsonPrimitive?.contentOrNull == "quota_exceeded",
            )
        }
        val receipt = json["receipt"]?.takeIf { it !is JsonNull }?.jsonObject
            ?: throw AiScanError(AI_SCAN_EMPTY)

        return aiReceiptDraft(
            receipt = receipt.toAiReceipt(),
            currencyHint = currencyHint,
            rawText = rawText,
        )
    }

    suspend fun saveScan(input: SaveScanInput): String {
        val id = UUID.randomUUID().toString()
        val ts = Instant.now().toString()
        val userId = getUserId()
        val rawTextCapped = input.rawText?.take(8000)

        db.execute(
            """INSERT INTO receipt_scans (
                id, user_id, source, engine, merchant, occurred_at, currency,
                subtotal, tax, service_charge, tip, discount, total, confidence,
                raw_text, parsed_json, transaction_id, expense_id, image_path,
                created_at, updated_at
               ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
            listOf(
                id, userId, input.source, input.engine, input.merchant, input.occurredAt, input.currency,
                input.subtotal, input.tax, input.serviceCharge, input.tip, input.discount, input.total, input.confidence,
                rawTextCapped, input.parsedJson, null, null, null,
                ts, ts
            )
        )
        return id
    }

    suspend fun updateScanDraft(scanId: String, input: UpdateScanDraftInput) {
        val ts = Instant.now().toString()
        db.execute(
            """UPDATE receipt_scans SET
                engine = ?, merchant = ?, occurred_at = ?, currency = ?,
                subtotal = ?, tax = ?, service_charge = ?, tip = ?, discount = ?,
                total = ?, confidence = ?, parsed_json = ?, updated_at = ?
               WHERE id = ? AND deleted_at IS NULL""",
            listOf(
                input.engine, input.merchant, input.occurredAt, input.currency,
                input.subtotal, input.tax, input.serviceCharge, input.tip, input.discount,
                input.total, input.confidence, input.parsedJson, ts,
                scanId
            )
        )
    }

    suspend fun linkScan(scanId: String, transactionId: String? = null, expenseId: String? = null) {
        val ts = Instant.now().toString()
        val sets = mutableListOf<String>()
        val params = mutableListOf<Any?>()

        if (transactionId != null) {
            sets.add("transaction_id = ?")
            params.add(transactionId)
        }
        if (expenseId != null) {
            sets.add("expense_id = ?")
            params.add(expenseId)
        }
        if (sets.isEmpty()) return

        sets.add("updated_at = ?")
        params.add(ts)
        params.add(scanId)

        db.execute(
            "UPDATE receipt_scans SET ${sets.joinToString(", ")} WHERE id = ? AND deleted_at IS NULL",
            params
        )
    }

    suspend fun get(scanId: String): ReceiptScanRow? = db.getOptional(
        sql = "SELECT * FROM receipt_scans WHERE id = ? AND deleted_at IS NULL",
        parameters = listOf(scanId),
        mapper = { cursor ->
            ReceiptScanRow(
                id = cursor.getString("id"),
                userId = cursor.getString("user_id"),
                source = cursor.getString("source"),
                engine = cursor.getString("engine"),
                merchant = cursor.getStringOptional("merchant"),
                occurredAt = cursor.getStringOptional("occurred_at"),
                currency = cursor.getStringOptional("currency"),
                subtotal = cursor.getLongOptional("subtotal"),
                tax = cursor.getLongOptional("tax"),
                serviceCharge = cursor.getLongOptional("service_charge"),
                tip = cursor.getLongOptional("tip"),
                discount = cursor.getLongOptional("discount"),
                total = cursor.getLongOptional("total"),
                confidence = cursor.getLongOptional("confidence"),
                rawText = cursor.getStringOptional("raw_text"),
                parsedJson = cursor.getStringOptional("parsed_json"),
                transactionId = cursor.getStringOptional("transaction_id"),
                expenseId = cursor.getStringOptional("expense_id"),
                imagePath = cursor.getStringOptional("image_path"),
                createdAt = cursor.getString("created_at"),
                updatedAt = cursor.getString("updated_at")
            )
        }
    )

    suspend fun list(limit: Int = 50): List<ReceiptScanRow> = db.getAll(
        sql = "SELECT * FROM receipt_scans WHERE deleted_at IS NULL ORDER BY created_at DESC LIMIT ?",
        parameters = listOf(limit),
        mapper = { cursor ->
            ReceiptScanRow(
                id = cursor.getString("id"),
                userId = cursor.getString("user_id"),
                source = cursor.getString("source"),
                engine = cursor.getString("engine"),
                merchant = cursor.getStringOptional("merchant"),
                occurredAt = cursor.getStringOptional("occurred_at"),
                currency = cursor.getStringOptional("currency"),
                subtotal = cursor.getLongOptional("subtotal"),
                tax = cursor.getLongOptional("tax"),
                serviceCharge = cursor.getLongOptional("service_charge"),
                tip = cursor.getLongOptional("tip"),
                discount = cursor.getLongOptional("discount"),
                total = cursor.getLongOptional("total"),
                confidence = cursor.getLongOptional("confidence"),
                rawText = cursor.getStringOptional("raw_text"),
                parsedJson = cursor.getStringOptional("parsed_json"),
                transactionId = cursor.getStringOptional("transaction_id"),
                expenseId = cursor.getStringOptional("expense_id"),
                imagePath = cursor.getStringOptional("image_path"),
                createdAt = cursor.getString("created_at"),
                updatedAt = cursor.getString("updated_at")
            )
        }
    )

    suspend fun delete(scanId: String) {
        val ts = Instant.now().toString()
        db.execute(
            "UPDATE receipt_scans SET deleted_at = ?, updated_at = ? WHERE id = ?",
            listOf(ts, ts, scanId)
        )
    }
}

/**
 * A failed AI read.
 *
 * [quotaExceeded] is separate because it is the only failure the user can do
 * something about: web shows the upgrade path for it and a plain message for
 * everything else.
 */
class AiScanError(message: String, val quotaExceeded: Boolean = false) : Exception(message)

private const val AI_SCAN_FAILURE = "Couldn't reach the scanner. Check your connection."
private const val AI_SCAN_EMPTY = "The scan came back empty. Try a clearer photo."

/**
 * The edge function's `receipt` object, as Domain's input type.
 *
 * Deliberately tolerant: every field is optional and a wrong TYPE reads as
 * absent rather than throwing. A model reply is untrusted input, and a strict
 * decoder here would turn one odd field into a total failure where web would
 * have shown the user a partial draft they could fix.
 */
private fun JsonObject.aiText(key: String): String? =
    this[key]?.let { if (it is JsonPrimitive && it !is JsonNull) it.content else null }

/**
 * A JSON NUMBER, or null.
 *
 * `isString` is checked because web's guard is `typeof raw.amount !== "number"`
 * -- a model that returns `"12.50"` as a string is REJECTED there, and the line
 * dropped. Accepting it here would make the two clients disagree about whether
 * a bill has that line on it at all.
 */
private fun JsonObject.aiNumber(key: String): Double? =
    this[key]?.let { if (it is JsonPrimitive && !it.isString) it.doubleOrNull else null }

private fun JsonObject.toAiReceipt(): AiReceipt = AiReceipt(
    merchant = aiText("merchant"),
    date = aiText("date"),
    currency = aiText("currency"),
    total = aiNumber("total"),
    confidence = aiNumber("confidence"),
    lines = this["lines"]?.let { element ->
        runCatching { element.jsonArray }.getOrNull()?.mapNotNull { line ->
            runCatching { line.jsonObject }.getOrNull()?.let { o ->
                AiLine(
                    kind = o.aiText("kind"),
                    description = o.aiText("description"),
                    quantity = o.aiNumber("quantity"),
                    unit = o.aiText("unit"),
                    // The wire name is snake_case; Domain's field is not.
                    // `receipts-ai.json` carries the WIRE name so the corpus
                    // pins it and both platforms' registrations must agree.
                    unitPrice = o.aiNumber("unit_price"),
                    amount = o.aiNumber("amount"),
                )
            }
        }
    } ?: emptyList(),
)
